### PR TITLE
[Merged by Bors] - feat: port Order.Partition.Finpartition

### PR DESCRIPTION
**Pull request by mo271**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by nomeata** [comment]
Fixed the proofs. Needed to add an instance to `Finset/Basic.lean`; reviewers please take a look.

**Comment by mo271** [comment]
> All done but the very last lemma, there a two-line `simp only` invocation no longer does what it used to do.

I think I got the first simp only invocation right now, but the proof is still not finished...

**Comment by nomeata** [comment]
After a good night sleep I could solve it. My approach was to edit the lean3 proof first and break the `simp only` into a sequence of `simp_rw` steps. This way I found out which lemmas needed to apply in which order (and some were not needed at all), and that sequence of `simp_rw` steps worked in lean4 as well.

File: Mathlib/Order/Partition/Finpartition.lean [hunk 1]
```diff
@@ -0,0 +1,618 @@
...
+  simp_rw [mem_image]
+  simp_rw [mem_powerset]
+  simp_rw [mem_filter]
+  simp_rw [and_imp]
+  simp_rw [Finset.Nonempty]
+  simp_rw [exists_imp]
+  simp_rw [mem_atomise]
+  simp_rw [and_imp]
+  simp_rw [Finset.Nonempty]
+  simp_rw [exists_imp]
```
Comment (jcommelin) [inline Mathlib/Order/Partition/Finpartition.lean:610]:
```suggestion
  simp_rw [mem_image, mem_powerset, mem_filter, and_imp, Finset.Nonempty,
    exists_imp, mem_atomise, and_imp,Finset.Nonempty, exists_imp]
```

**Review by jcommelin — COMMENTED** [review]
lgtm

bors d+

File: Mathlib/Order/Partition/Finpartition.lean (hunk 1 above)
Comment (nomeata) [reply Mathlib/Order/Partition/Finpartition.lean]: Hmm, that loses the ability to look at the individual steps easily (wasn’t that different in lean3? Or only with `rw`), but it works.

**Comment by nomeata** [comment]
@mo271 I applied the suggested changes, so I think you can merge

**Comment by nomeata** [comment]
I get permission denied from bots trying to see the log. Maybe we need to merge master and see what broke

**Comment by mo271** [comment]
Trying to merge manually with master, because bors attempt to merge failed...

**Comment by mo271** [comment]
Ah, this seems to be about the BigOperators, will send a fix...

**Pull request closed without merging**